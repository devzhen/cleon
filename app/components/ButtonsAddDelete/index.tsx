import { Button, Container } from "./styled";

type ButtonAddDeleteProps = {
  add: () => void;
  remove: () => void;
  marginTop?: number;
  marginRight?: number;
};

const ButtonAddDelete = (props: ButtonAddDeleteProps) => {
  const { add, remove, marginRight, marginTop } = props;

  return (
    <Container $marginRight={marginRight} $marginTop={marginTop}>
      <Button onClick={add}>+</Button>
      <Button onClick={remove} $marginTop={10}>
        -
      </Button>
    </Container>
  );
};

export default ButtonAddDelete;
