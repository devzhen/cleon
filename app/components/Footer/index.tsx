import Image from 'next/image';

import styles from './Footer.module.css';

export default function Footer() {
  console.log(111, process.env.NEXT_PUBLIC_GITHUB_REPO);
  console.log(222, process.env.GITHUB_REPO);

  return (
    <footer className={styles.footer}>
      <div className={styles.footerBg}>
        <a href={process.env.NEXT_PUBLIC_GITHUB_REPO}>
          <Image src="/github-mark.png" width={24} height={24} alt="Img" draggable={false} />
          <Image src="/github-logo.png" width={54} height={24} alt="Img" draggable={false} />
          <span>go to the repo</span>
        </a>
      </div>
    </footer>
  );
}
