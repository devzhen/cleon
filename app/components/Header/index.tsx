import styles from './Header.module.css';

export default function Header() {
  return (
    <header className={styles.header}>
      <div />
      <div>
        <div className={styles['header-logo']}>
          <div />
          <div className={styles['header-logo-title']}>
            <span>cl√àon</span>
          </div>
          <div className={styles['header-logo-subtitle']}>
            <span>l`usine incontournable pour l`automobile de demain</span>
          </div>
        </div>
      </div>
    </header>
  );
}
