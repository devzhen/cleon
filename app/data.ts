import { v4 as uuid } from "uuid";

export const stickers = [
  {
    id: uuid(),
    top: 110,
    left: 230,
    createdAt: new Date().toISOString(),
    text: ''
  }
];
