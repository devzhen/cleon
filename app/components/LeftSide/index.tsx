import Image from "next/image";
import { useRef } from "react";

import { BOARD_TYPE } from "@/app/constants";
import useDraggableAreas from "@/app/hooks/useDraggableAreas";
import useSticker from "@/app/hooks/useSticker";
import getCoords from "@/app/utils/getCoords";
import isPointInPath from "@/app/utils/isPointInPath";

import ButtonsAddDelete from "../ButtonsAddDelete";
import ModalAddSticker from "../ModalAddSticker";
import StickerComponent from "../Sticker";

import {
  Container,
  Cell,
  FloatingWrapper,
  Banner,
  LogoWrapper,
} from "./styled";

export default function LeftSide() {
  // Current draggable sticker
  const currentDragElementRef = useRef<HTMLDivElement>();

  // Current draggable sticker's parent
  const currentDragElementParentRef = useRef<HTMLDivElement>();

  // Drag containers
  const dragContainerRef = useRef<HTMLDivElement>(null);
  const dragArea1Ref = useRef<HTMLDivElement>(null);
  const dragArea2Ref = useRef<HTMLDivElement>(null);
  const dragArea3Ref = useRef<HTMLDivElement>(null);

  // Banners, logo, buttons
  const leftBannerWrapperRef = useRef<HTMLDivElement>(null);
  const logoWrapperRef = useRef<HTMLDivElement>(null);
  const buttonsWrapperRef = useRef<HTMLDivElement>(null);
  const lastBannerWrapperRef = useRef<HTMLDivElement>(null);

  /**
   * Init draggable areas
   */
  const initDragAreas = () => {
    if (
      !dragContainerRef.current ||
      !dragArea1Ref.current ||
      !dragArea2Ref.current ||
      !dragArea3Ref.current ||
      !leftBannerWrapperRef.current ||
      !logoWrapperRef.current ||
      !buttonsWrapperRef.current ||
      !lastBannerWrapperRef.current
    ) {
      return;
    }

    const dragContainerRect = dragContainerRef.current.getBoundingClientRect();

    const leftBannerWrapperRect =
      leftBannerWrapperRef.current.getBoundingClientRect();

    const logoWrapperRect = logoWrapperRef.current.getBoundingClientRect();

    const buttonsWrapperRect =
      buttonsWrapperRef.current.getBoundingClientRect();

    const lastBannerWrapperRect =
      lastBannerWrapperRef.current.getBoundingClientRect();

    const BORDER_WIDTH = 3;

    const dragArea1Rect = {
      left: leftBannerWrapperRect.width,
      top: logoWrapperRect.height,
      width:
        dragContainerRect.width -
        leftBannerWrapperRect.width -
        buttonsWrapperRect.width -
        BORDER_WIDTH * 2,
      height:
        dragContainerRect.height -
        (logoWrapperRect.height +
          lastBannerWrapperRect.height +
          BORDER_WIDTH * 2),
    };

    dragArea1Ref.current.style.top = `${dragArea1Rect.top}px`;
    dragArea1Ref.current.style.left = `${dragArea1Rect.left}px`;
    dragArea1Ref.current.style.width = `${dragArea1Rect.width}px`;
    dragArea1Ref.current.style.height = `${dragArea1Rect.height}px`;

    const dragArea2Rect = {
      left: leftBannerWrapperRect.width,
      top: buttonsWrapperRect.height,
      width:
        dragContainerRect.width -
        leftBannerWrapperRect.width -
        BORDER_WIDTH * 2,
      height:
        dragContainerRect.height -
        (buttonsWrapperRect.height +
          lastBannerWrapperRect.height +
          BORDER_WIDTH * 2),
    };

    dragArea2Ref.current.style.top = `${dragArea2Rect.top}px`;
    dragArea2Ref.current.style.left = `${dragArea2Rect.left}px`;
    dragArea2Ref.current.style.width = `${dragArea2Rect.width}px`;
    dragArea2Ref.current.style.height = `${dragArea2Rect.height}px`;

    const dragArea3Rect = {
      left: 0,
      top: leftBannerWrapperRect.height,
      width:
        dragContainerRect.width -
        lastBannerWrapperRect.width -
        BORDER_WIDTH * 2,
      height:
        dragContainerRect.height -
        (leftBannerWrapperRect.height + BORDER_WIDTH * 2),
    };

    dragArea3Ref.current.style.top = `${dragArea3Rect.top}px`;
    dragArea3Ref.current.style.left = `${dragArea3Rect.left}px`;
    dragArea3Ref.current.style.width = `${dragArea3Rect.width}px`;
    dragArea3Ref.current.style.height = `${dragArea3Rect.height}px`;

    currentDragElementParentRef.current = dragArea2Ref.current;
  };

  /**
   * Detect a sticker's parent container
   */
  const detectParentContainer = () => {
    if (
      !dragArea1Ref.current ||
      !dragArea2Ref.current ||
      !dragArea3Ref.current ||
      !leftBannerWrapperRef.current ||
      !logoWrapperRef.current ||
      !buttonsWrapperRef.current ||
      !lastBannerWrapperRef.current
    ) {
      return;
    }

    const stickerCoords = getCoords(
      currentDragElementRef.current as HTMLDivElement
    );

    const dragArea1Coords = getCoords(dragArea1Ref.current);
    const dragArea2Coords = getCoords(dragArea2Ref.current);
    const dragArea3Coords = getCoords(dragArea3Ref.current);

    // The drag area 1
    if (currentDragElementParentRef.current === dragArea1Ref.current) {
      // Attach to the drag area 2
      if (
        stickerCoords.top > dragArea2Coords.top &&
        !isPointInPath(
          leftBannerWrapperRef.current,
          stickerCoords.left,
          stickerCoords.top
        ) &&
        !isPointInPath(
          buttonsWrapperRef.current,
          stickerCoords.right,
          stickerCoords.top
        )
      ) {
        currentDragElementParentRef.current = dragArea2Ref.current;
      }
    }

    // The drag area 2
    if (currentDragElementParentRef.current === dragArea2Ref.current) {
      // Attach to the drag area 3
      if (
        stickerCoords.top > dragArea3Coords.top &&
        stickerCoords.left > dragArea3Coords.left &&
        stickerCoords.right < dragArea3Coords.right &&
        stickerCoords.bottom < dragArea3Coords.bottom
      ) {
        currentDragElementParentRef.current = dragArea3Ref.current;
      }

      // Attach to the drag area 1
      if (
        stickerCoords.top < dragArea2Coords.top &&
        stickerCoords.right <= dragArea1Coords.right
      ) {
        currentDragElementParentRef.current = dragArea1Ref.current;
      }
    }

    // The drag area 3
    if (currentDragElementParentRef.current === dragArea3Ref.current) {
      // If move to top, attach to the drag area 2
      if (
        stickerCoords.top < dragArea3Coords.top &&
        stickerCoords.left > dragArea3Coords.left &&
        stickerCoords.right < dragArea3Coords.right &&
        stickerCoords.bottom < dragArea3Coords.bottom &&
        !isPointInPath(
          leftBannerWrapperRef.current,
          stickerCoords.left,
          stickerCoords.top
        )
      ) {
        currentDragElementParentRef.current = dragArea2Ref.current;
      }

      // If move to right, attach to the drag area 2
      if (
        stickerCoords.top > dragArea3Coords.top &&
        stickerCoords.left > dragArea3Coords.left &&
        stickerCoords.right > dragArea3Coords.right &&
        stickerCoords.bottom < dragArea3Coords.bottom &&
        !isPointInPath(
          lastBannerWrapperRef.current,
          stickerCoords.right,
          stickerCoords.bottom
        )
      ) {
        currentDragElementParentRef.current = dragArea2Ref.current;
      }

      // If move to right, attach to the drag area 2
      if (
        stickerCoords.top <= dragArea3Coords.top &&
        stickerCoords.left > dragArea3Coords.left &&
        stickerCoords.right >= dragArea3Coords.right &&
        stickerCoords.bottom < dragArea3Coords.bottom &&
        isPointInPath(
          lastBannerWrapperRef.current,
          stickerCoords.right,
          stickerCoords.bottom
        )
      ) {
        currentDragElementParentRef.current = dragArea2Ref.current;
      }
    }
  };

  /**
   * Correct a sticker's position relative to a parent
   */
  const correctPositionRelativeToParent = () => {
    if (
      !currentDragElementRef.current ||
      !currentDragElementParentRef.current
    ) {
      return;
    }

    const stickerCoords = getCoords(
      currentDragElementRef.current as HTMLDivElement
    );

    const parentCoords = getCoords(
      currentDragElementParentRef.current as HTMLDivElement
    );

    if (stickerCoords.left < parentCoords.left) {
      currentDragElementRef.current.style.left = `${currentDragElementParentRef.current.offsetLeft}px`;
    }

    if (stickerCoords.top < parentCoords.top) {
      currentDragElementRef.current.style.top = `${currentDragElementParentRef.current.offsetTop}px`;
    }

    if (stickerCoords.left + stickerCoords.width > parentCoords.right) {
      currentDragElementRef.current.style.left = `${
        currentDragElementParentRef.current.offsetLeft +
        parentCoords.width -
        stickerCoords.width
      }px`;
    }

    if (stickerCoords.top + stickerCoords.width > parentCoords.bottom) {
      currentDragElementRef.current.style.top = `${
        currentDragElementParentRef.current?.offsetTop +
        parentCoords.height -
        stickerCoords.height
      }px`;
    }
  };

  /**
   * Correct a sticker's position relative to a drag container
   */
  const correctPositionRelativeToDragContainer = () => {
    if (!currentDragElementRef.current || !dragContainerRef.current) {
      return;
    }

    const stickerCoords = getCoords(
      currentDragElementRef.current as HTMLDivElement
    );

    const dragContainerCoords = getCoords(
      dragContainerRef.current as HTMLDivElement
    );

    if (stickerCoords.left < dragContainerCoords.left) {
      currentDragElementRef.current.style.left = `${0}px`;
    }

    if (stickerCoords.top < dragContainerCoords.top) {
      currentDragElementRef.current.style.top = `${0}px`;
    }

    if (stickerCoords.left + stickerCoords.width > dragContainerCoords.right) {
      currentDragElementRef.current.style.left = `${
        dragContainerCoords.width - stickerCoords.width
      }px`;
    }

    if (stickerCoords.top + stickerCoords.width > dragContainerCoords.bottom) {
      currentDragElementRef.current.style.top = `${
        dragContainerCoords.height - stickerCoords.height
      }px`;
    }
  };

  /**
   * Use draggable areas hook
   */
  const { onMouseDown, onMouseMove } = useDraggableAreas({
    currentDragElementRef,
    dragContainerRef,
    onMountHandler: initDragAreas,
    onMouseMoveHandler: () => {
      detectParentContainer();

      correctPositionRelativeToParent();

      correctPositionRelativeToDragContainer();
    },
    onMouseDownHandler: () => {},
  });

  /**
   * Use sticker hook
   */
  const {
    closeModalHandler,
    createSticker,
    editSticker,
    editedSticker,
    isModalOpen,
    setModalVisibility,
    stickers,
    isDeleteMode,
    toggleDeleteMode,
    removeSticker,
  } = useSticker({
    board: BOARD_TYPE.left,
  });

  return (
    <>
      <Container
        className="sticker-container"
        onMouseMove={onMouseMove as VoidFunction}
        ref={dragContainerRef}
      >
        <div className="drag-area" ref={dragArea1Ref} />
        <div className="drag-area" ref={dragArea2Ref} />
        <div className="drag-area" ref={dragArea3Ref} />
        <Cell>
          <FloatingWrapper
            className="floating-wrapper"
            $paddingTop={80}
            ref={leftBannerWrapperRef}
          >
            <Banner $marginLeft={10}>
              <Image
                src="/banner1.jpeg"
                width={185}
                height={132}
                alt="Banner 1"
                priority
                draggable={false}
              />
            </Banner>
            <Banner $marginTop={10} $marginLeft={10}>
              <Image
                src="/banner2.jpeg"
                width={185}
                height={132}
                alt="Banner 2"
                draggable={false}
                priority
              />
            </Banner>
            <Banner $marginTop={10} $marginLeft={10}>
              <Image
                src="/banner3.jpeg"
                width={185}
                height={132}
                alt="Banner 3"
                draggable={false}
                priority
              />
            </Banner>
          </FloatingWrapper>
        </Cell>
        <Cell $justifyContent="center" $alignItems="start">
          <FloatingWrapper className="floating-wrapper" ref={logoWrapperRef}>
            <LogoWrapper>
              <Image
                src="/gerb-1.png"
                width={39}
                height={52}
                alt="Picture of the author"
                draggable={false}
              />
              <span>sÈcuritÈ</span>
            </LogoWrapper>
          </FloatingWrapper>
        </Cell>
        <Cell $alignItems="start" $justifyContent="flex-end">
          <FloatingWrapper className="floating-wrapper" ref={buttonsWrapperRef}>
            <ButtonsAddDelete
              add={setModalVisibility(true)}
              remove={toggleDeleteMode}
              marginTop={22}
              marginRight={16}
            />
          </FloatingWrapper>
        </Cell>
        <Cell />
        <Cell />
        <Cell />
        <Cell />
        <Cell />
        <Cell $alignItems="end" $justifyContent="flex-end">
          <FloatingWrapper
            className="floating-wrapper"
            ref={lastBannerWrapperRef}
          >
            <Banner
              $marginRight={10}
              $marginBottom={10}
              $width={129}
              $height={188}
            >
              <Image
                src="/banner4.jpeg"
                width={129}
                height={188}
                alt="Banner 4"
                draggable={false}
              />
            </Banner>
          </FloatingWrapper>
        </Cell>
        {stickers.map((item) => (
          <StickerComponent
            onMouseDown={onMouseDown as VoidFunction}
            key={item.id}
            sticker={item}
            editSticker={editSticker(item)}
            isDeleteMode={isDeleteMode}
            removeSticker={removeSticker(item.id)}
          />
        ))}
      </Container>
      {isModalOpen && (
        <ModalAddSticker
          closeModal={closeModalHandler}
          createSticker={createSticker}
          board={BOARD_TYPE.left}
          sticker={editedSticker}
        />
      )}
    </>
  );
}
