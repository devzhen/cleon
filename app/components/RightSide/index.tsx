import Image from "next/image";
import { useRef } from "react";

import { BOARD_TYPE } from "@/app/constants";
import useDraggableAreas from "@/app/hooks/useDraggableAreas";
import useSticker from "@/app/hooks/useSticker";
import getCoords from "@/app/utils/getCoords";
import isPointInPath from "@/app/utils/isPointInPath";

import ButtonsAddDelete from "../ButtonsAddDelete";
import ModalAddSticker from "../ModalAddSticker";
import StickerComponent from "../Sticker";

import { Container, Header, ButtonsRow, BannersRow } from "./styled";

export default function LeftSide() {
  // Current draggable sticker
  const currentDragElementRef = useRef<HTMLDivElement>();

  // Current draggable sticker's parent
  const currentDragElementParentRef = useRef<HTMLDivElement>();

  // Drag containers
  const dragContainerRef = useRef<HTMLDivElement>(null);
  const dragArea1Ref = useRef<HTMLDivElement>(null);
  const dragArea2Ref = useRef<HTMLDivElement>(null);
  const dragArea3Ref = useRef<HTMLDivElement>(null);
  const dragArea4Ref = useRef<HTMLDivElement>(null);
  const dragArea5Ref = useRef<HTMLDivElement>(null);

  // Banners, logos, images
  const headerRef = useRef<HTMLDivElement>(null);
  const buttonAddDeleteRef = useRef<HTMLDivElement>(null);
  const topImagesRef = useRef<HTMLImageElement>(null);
  const logoRef = useRef<HTMLImageElement>(null);
  const bottomLeftImagesRef = useRef<HTMLDivElement>(null);
  const bottomRightImagesRef = useRef<HTMLDivElement>(null);

  /**
   * Init draggable areas
   */
  const initDragAreas = () => {
    if (
      !dragContainerRef.current ||
      !dragArea1Ref.current ||
      !dragArea2Ref.current ||
      !dragArea3Ref.current ||
      !dragArea4Ref.current ||
      !dragArea5Ref.current ||
      !headerRef.current ||
      !logoRef.current ||
      !buttonAddDeleteRef.current ||
      !topImagesRef.current ||
      !bottomLeftImagesRef.current ||
      !bottomRightImagesRef.current
    ) {
      return;
    }

    const BORDER_WIDTH = 3;

    const dragContainerRefRect =
      dragContainerRef.current.getBoundingClientRect();
    const headerRect = headerRef.current.getBoundingClientRect();
    const topImagesRefRect = topImagesRef.current.getBoundingClientRect();
    const buttonAddDeleteRefRect =
      buttonAddDeleteRef.current.getBoundingClientRect();
    const bottomLeftImagesRefRect =
      bottomLeftImagesRef.current.getBoundingClientRect();
    const bottomRightImagesRefRect =
      bottomRightImagesRef.current.getBoundingClientRect();

    const dragArea1Rect = {
      left: 0,
      top: 0,
      width: logoRef.current.offsetLeft,
      height: bottomLeftImagesRef.current.offsetTop,
    };

    dragArea1Ref.current.style.top = `${dragArea1Rect.top}px`;
    dragArea1Ref.current.style.left = `${dragArea1Rect.left}px`;
    dragArea1Ref.current.style.width = `${dragArea1Rect.width}px`;
    dragArea1Ref.current.style.height = `${dragArea1Rect.height}px`;
    dragArea1Ref.current.style.background = "blue";
    dragArea1Ref.current.style.opacity = "0.5";

    const dragArea2Rect = {
      left: 0,
      top: headerRect.height,
      width: buttonAddDeleteRef.current.offsetLeft,
      height: bottomLeftImagesRef.current.offsetTop - headerRect.height,
    };

    dragArea2Ref.current.style.top = `${dragArea2Rect.top}px`;
    dragArea2Ref.current.style.left = `${dragArea2Rect.left}px`;
    dragArea2Ref.current.style.width = `${dragArea2Rect.width}px`;
    dragArea2Ref.current.style.height = `${dragArea2Rect.height}px`;
    dragArea2Ref.current.style.opacity = "0.5";

    const dragArea3Rect = {
      left: 0,
      top: buttonAddDeleteRef.current.offsetTop + buttonAddDeleteRefRect.height,
      width: topImagesRef.current.offsetLeft,
      height:
        bottomLeftImagesRef.current.offsetTop -
        headerRect.height -
        buttonAddDeleteRefRect.height,
    };

    dragArea3Ref.current.style.top = `${dragArea3Rect.top}px`;
    dragArea3Ref.current.style.left = `${dragArea3Rect.left}px`;
    dragArea3Ref.current.style.width = `${dragArea3Rect.width}px`;
    dragArea3Ref.current.style.height = `${dragArea3Rect.height}px`;
    dragArea3Ref.current.style.opacity = "0.5";

    const dragArea4Rect = {
      left: 0,
      top: topImagesRef.current.offsetTop + topImagesRefRect.height,
      width: dragContainerRefRect.width - BORDER_WIDTH * 2,
      height:
        bottomRightImagesRef.current.offsetTop -
        headerRect.height -
        topImagesRefRect.height,
    };

    dragArea4Ref.current.style.top = `${dragArea4Rect.top}px`;
    dragArea4Ref.current.style.left = `${dragArea4Rect.left}px`;
    dragArea4Ref.current.style.width = `${dragArea4Rect.width}px`;
    dragArea4Ref.current.style.height = `${dragArea4Rect.height}px`;
    dragArea4Ref.current.style.opacity = "0.5";

    const dragArea5Rect = {
      left:
        bottomLeftImagesRefRect.width + bottomLeftImagesRef.current.offsetLeft,
      top:
        bottomLeftImagesRef.current.offsetTop - bottomLeftImagesRefRect.height,
      width:
        bottomRightImagesRef.current.offsetLeft -
        bottomLeftImagesRefRect.width -
        bottomLeftImagesRef.current.offsetLeft,
      height: bottomRightImagesRefRect.height * 2,
    };

    dragArea5Ref.current.style.top = `${dragArea5Rect.top}px`;
    dragArea5Ref.current.style.left = `${dragArea5Rect.left}px`;
    dragArea5Ref.current.style.width = `${dragArea5Rect.width}px`;
    dragArea5Ref.current.style.height = `${dragArea5Rect.height}px`;
    dragArea5Ref.current.style.opacity = "0.5";
  };

  /**
   * Detect parent container
   */
  const detectParentContainer = () => {
    if (
      !dragContainerRef.current ||
      !dragArea1Ref.current ||
      !dragArea2Ref.current ||
      !dragArea3Ref.current ||
      !dragArea4Ref.current ||
      !dragArea5Ref.current ||
      !headerRef.current ||
      !logoRef.current ||
      !buttonAddDeleteRef.current ||
      !topImagesRef.current ||
      !bottomLeftImagesRef.current ||
      !bottomRightImagesRef.current
    ) {
      return;
    }

    const stickerCoords = getCoords(
      currentDragElementRef.current as HTMLDivElement
    );

    const dragArea1Coords = getCoords(dragArea1Ref.current);
    const dragArea2Coords = getCoords(dragArea2Ref.current);
    const dragArea3Coords = getCoords(dragArea3Ref.current);
    const dragArea4Coords = getCoords(dragArea4Ref.current);
    const dragArea5Coords = getCoords(dragArea5Ref.current);

    // The drag area 1
    if (currentDragElementParentRef.current === dragArea1Ref.current) {
      if (
        stickerCoords.top > dragArea1Coords.top &&
        isPointInPath(
          dragArea2Ref.current,
          stickerCoords.left,
          stickerCoords.top
        )
      ) {
        currentDragElementParentRef.current = dragArea2Ref.current;
      }
    }

    // The drag area 2
    if (currentDragElementParentRef.current === dragArea2Ref.current) {
      if (
        stickerCoords.top < dragArea2Coords.top &&
        isPointInPath(
          dragArea1Ref.current,
          stickerCoords.right,
          stickerCoords.top
        )
      ) {
        currentDragElementParentRef.current = dragArea1Ref.current;
      }

      if (
        stickerCoords.top >= dragArea3Coords.top &&
        isPointInPath(
          dragArea3Ref.current,
          stickerCoords.right,
          stickerCoords.top
        )
      ) {
        currentDragElementParentRef.current = dragArea3Ref.current;
      }
    }

    // The drag area 3
    if (currentDragElementParentRef.current === dragArea3Ref.current) {
      if (
        stickerCoords.top < dragArea3Coords.top &&
        isPointInPath(
          dragArea2Ref.current,
          stickerCoords.right,
          stickerCoords.top
        )
      ) {
        currentDragElementParentRef.current = dragArea2Ref.current;
      }

      if (
        isPointInPath(
          dragArea4Ref.current,
          stickerCoords.right,
          stickerCoords.top
        )
      ) {
        currentDragElementParentRef.current = dragArea4Ref.current;
      }
    }

    // The drag area 4
    if (currentDragElementParentRef.current === dragArea4Ref.current) {
      if (
        stickerCoords.top < dragArea4Coords.top &&
        isPointInPath(
          dragArea3Ref.current,
          stickerCoords.right,
          stickerCoords.top
        )
      ) {
        currentDragElementParentRef.current = dragArea3Ref.current;
      }

      if (
        isPointInPath(
          dragArea5Ref.current,
          stickerCoords.left,
          stickerCoords.top
        ) &&
        isPointInPath(
          dragArea5Ref.current,
          stickerCoords.right,
          stickerCoords.top
        ) &&
        isPointInPath(
          dragArea5Ref.current,
          stickerCoords.left,
          stickerCoords.bottom
        ) &&
        isPointInPath(
          dragArea5Ref.current,
          stickerCoords.right,
          stickerCoords.bottom
        )
      ) {
        currentDragElementParentRef.current = dragArea5Ref.current;
      }
    }

    // The drag area 5
    if (currentDragElementParentRef.current === dragArea5Ref.current) {
      if (
        stickerCoords.top < dragArea5Coords.top &&
        (!isPointInPath(
          dragArea5Ref.current,
          stickerCoords.left,
          stickerCoords.top
        ) ||
          !isPointInPath(
            dragArea5Ref.current,
            stickerCoords.right,
            stickerCoords.top
          ))
      ) {
        currentDragElementParentRef.current = dragArea4Ref.current;
      }
    }
  };

  /**
   * Correct a sticker's position relative to a parent
   */
  const correctPositionRelativeToParent = () => {
    if (
      !currentDragElementRef.current ||
      !currentDragElementParentRef.current
    ) {
      return;
    }

    const stickerCoords = getCoords(
      currentDragElementRef.current as HTMLDivElement
    );

    const parentCoords = getCoords(
      currentDragElementParentRef.current as HTMLDivElement
    );

    if (stickerCoords.left < parentCoords.left) {
      currentDragElementRef.current.style.left = `${currentDragElementParentRef.current.offsetLeft}px`;
    }

    if (stickerCoords.top < parentCoords.top) {
      currentDragElementRef.current.style.top = `${currentDragElementParentRef.current.offsetTop}px`;
    }

    if (stickerCoords.left + stickerCoords.width > parentCoords.right) {
      currentDragElementRef.current.style.left = `${
        currentDragElementParentRef.current.offsetLeft +
        parentCoords.width -
        stickerCoords.width
      }px`;
    }

    if (stickerCoords.top + stickerCoords.width > parentCoords.bottom) {
      currentDragElementRef.current.style.top = `${
        currentDragElementParentRef.current?.offsetTop +
        parentCoords.height -
        stickerCoords.height
      }px`;
    }
  };

  /**
   * Correct a sticker's position relative to a drag container
   */
  const correctPositionRelativeToDragContainer = () => {
    if (!currentDragElementRef.current || !dragContainerRef.current) {
      return;
    }

    const stickerCoords = getCoords(
      currentDragElementRef.current as HTMLDivElement
    );

    const dragContainerCoords = getCoords(
      dragContainerRef.current as HTMLDivElement
    );

    if (stickerCoords.left < dragContainerCoords.left) {
      currentDragElementRef.current.style.left = `${0}px`;
    }

    if (stickerCoords.top < dragContainerCoords.top) {
      currentDragElementRef.current.style.top = `${0}px`;
    }

    if (stickerCoords.left + stickerCoords.width > dragContainerCoords.right) {
      currentDragElementRef.current.style.left = `${
        dragContainerCoords.width - stickerCoords.width
      }px`;
    }

    if (stickerCoords.top + stickerCoords.width > dragContainerCoords.bottom) {
      currentDragElementRef.current.style.top = `${
        dragContainerCoords.height - stickerCoords.height
      }px`;
    }
  };

  /**
   * Use draggable areas hook
   */
  const { onMouseDown, onMouseMove } = useDraggableAreas({
    currentDragElementRef,
    dragContainerRef,
    onMountHandler: () => {
      initDragAreas();

      currentDragElementParentRef.current =
        dragArea2Ref.current as HTMLDivElement;
    },
    onMouseMoveHandler: () => {
      detectParentContainer();
      correctPositionRelativeToParent();
      correctPositionRelativeToDragContainer();
    },
  });

  /**
   * Use sticker hook
   */
  const {
    closeModalHandler,
    createSticker,
    editSticker,
    editedSticker,
    isModalOpen,
    setModalVisibility,
    stickers,
  } = useSticker({
    board: BOARD_TYPE.right,
  });

  return (
    <>
      <Container
        className="sticker-container"
        ref={dragContainerRef}
        onMouseMove={onMouseMove as VoidFunction}
      >
        <div className="drag-area drag-area-1" ref={dragArea1Ref} />
        <div className="drag-area drag-area-2" ref={dragArea2Ref} />
        <div className="drag-area drag-area-3" ref={dragArea3Ref} />
        <div className="drag-area drag-area-4" ref={dragArea4Ref} />
        <div className="drag-area drag-area-5" ref={dragArea5Ref} />
        <Header ref={headerRef}>
          <Image
            src="/scrin-3-graf.png"
            width={47}
            height={47}
            alt="Picture of the author"
            draggable={false}
            ref={logoRef}
          />
          <span>vie de l’uet</span>
        </Header>
        <ButtonsRow>
          <ButtonsAddDelete
            add={setModalVisibility(true)}
            remove={() => {}}
            marginRight={10}
            ref={buttonAddDeleteRef}
          />
          <div ref={topImagesRef}>
            <Image
              src="/scrin3-a3-right1.jpg"
              width={262}
              height={185}
              alt="Picture of the author"
              draggable={false}
              priority
            />
            <Image
              src="/scrin3-a3-right2.jpg"
              width={262}
              height={185}
              alt="Picture of the author"
              draggable={false}
              priority
            />
          </div>
        </ButtonsRow>
        <div />
        <BannersRow>
          <div ref={bottomLeftImagesRef}>
            <Image
              src="/scrin3-a3-right3.jpg"
              width={262}
              height={185}
              alt="Picture of the author"
              draggable={false}
              priority
            />
            <Image
              src="/scrin3-a3-right4.jpg"
              width={262}
              height={185}
              alt="Picture of the author"
              draggable={false}
              priority
            />
          </div>
          <div ref={bottomRightImagesRef}>
            <Image
              src="/scrin3-a4-right1.jpg"
              width={129}
              height={188}
              alt="Picture of the author"
              draggable={false}
              priority
            />
            <Image
              src="/scrin3-a4-right2.jpg"
              width={129}
              height={188}
              alt="Picture of the author"
              draggable={false}
              priority
            />
          </div>
        </BannersRow>
        {stickers.map((item) => (
          <StickerComponent
            onMouseDown={onMouseDown as VoidFunction}
            key={item.id}
            sticker={item}
            editSticker={editSticker(item)}
          />
        ))}
      </Container>
      {isModalOpen && (
        <ModalAddSticker
          closeModal={closeModalHandler}
          createSticker={createSticker}
          board={BOARD_TYPE.left}
          sticker={editedSticker}
        />
      )}
    </>
  );
}
