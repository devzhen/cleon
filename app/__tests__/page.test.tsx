import { render } from '@testing-library/react';

import { LeftSideProps } from '../components/LeftSide';
import { RightSideProps } from '../components/RightSide';
import { stickers, firstId, secondId } from '../data';
import Page from '../page';

jest.mock('../api/stickers/fetchStickers', () =>
  jest.fn(() => ({
    left: {
      [firstId]: stickers[firstId],
    },
    right: {
      [secondId]: stickers[secondId],
    },
  })),
);

jest.mock(
  '../components/LeftSide',
  () =>
    function LeftSide(props: LeftSideProps) {
      return <div>{JSON.stringify(props)}</div>;
    },
);
jest.mock(
  '../components/RightSide',
  () =>
    function RightSide(props: RightSideProps) {
      return <div>{JSON.stringify(props)}</div>;
    },
);

describe('Page', () => {
  it('matches the snapshot', async () => {
    const { container } = render(await Page());

    expect(container).toMatchSnapshot();
  });
});
