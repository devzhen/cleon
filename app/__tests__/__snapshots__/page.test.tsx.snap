// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page matches the snapshot 1`] = `
<div>
  <main
    class="main className"
  >
    <header
      class="header"
    >
      <div />
      <div>
        <div
          class="header-logo"
        >
          <div />
          <div
            class="header-logo-title"
          >
            <span>
              cl√àon
            </span>
          </div>
          <div
            class="header-logo-subtitle"
          >
            <span>
              l\`usine incontournable pour l\`automobile de demain
            </span>
          </div>
        </div>
      </div>
    </header>
    <div>
      {"initialStickers":{"uuid/v4":{"id":"uuid/v4","top":449,"left":963,"createdAt":"2024-02-03T07:10:11.615Z","text":"Draggable element 4","board":"right","zIndex":3}}}
    </div>
    <div>
      {"initialStickers":{"uuid/v4":{"id":"uuid/v4","top":449,"left":963,"createdAt":"2024-02-03T07:10:11.615Z","text":"Draggable element 4","board":"right","zIndex":3}}}
    </div>
    <footer
      class="footer"
    >
      <div
        class="footerBg"
      >
        <a>
          <img
            alt="Img"
            data-nimg="1"
            decoding="async"
            draggable="false"
            height="24"
            loading="lazy"
            src="/_next/image?url=%2Fgithub-mark.png&w=48&q=75"
            srcset="/_next/image?url=%2Fgithub-mark.png&w=32&q=75 1x, /_next/image?url=%2Fgithub-mark.png&w=48&q=75 2x"
            style="color: transparent;"
            width="24"
          />
          <img
            alt="Img"
            data-nimg="1"
            decoding="async"
            draggable="false"
            height="24"
            loading="lazy"
            src="/_next/image?url=%2Fgithub-logo.png&w=128&q=75"
            srcset="/_next/image?url=%2Fgithub-logo.png&w=64&q=75 1x, /_next/image?url=%2Fgithub-logo.png&w=128&q=75 2x"
            style="color: transparent;"
            width="54"
          />
          <span>
            go to the repo
          </span>
        </a>
      </div>
    </footer>
  </main>
</div>
`;
